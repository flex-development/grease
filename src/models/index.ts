/**
 * @file Entry Point - Models
 * @module grease/models
 */

export { default as CommitGrammar } from './commit-grammar.model'
export { default as Commit } from './commit.model'
export { default as Version } from './version.model'
