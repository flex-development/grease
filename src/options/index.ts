/**
 * @file Entry Point - Options
 * @module grease/options
 */

export { default as GlobalOptions } from './global.options'
