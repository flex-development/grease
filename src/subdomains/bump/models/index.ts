/**
 * @file Entry Point - Models
 * @module grease/bump/models
 */

export { default as RecommendedBump } from './recommended-bump.model'
