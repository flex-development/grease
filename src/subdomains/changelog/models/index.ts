/**
 * @file Entry Point - Models
 * @module grease/changelog/models
 */

export {
  default as AbstractChangelogAggregator
} from './changelog-aggregator.abstract'
export { default as ChangelogAggregator } from './changelog-aggregator.model'
export {
  default as ChangelogEntry,
  type ChangelogEntryDTO
} from './changelog-entry.model'
export {
  default as AbstractChangelogFormatter
} from './changelog-formatter.abstract'
export { default as ChangelogFormatter } from './changelog-formatter.model'
export { default as ChangelogInfile } from './changelog-infile.model'
export {
  default as ChangelogStream,
  type ChangelogStreamDTO
} from './changelog-stream.model'
export { default as CommitGroup } from './commit-group.model'
export { default as CommitType } from './commit-type.model'
