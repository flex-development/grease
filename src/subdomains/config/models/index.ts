/**
 * @file Entry Point - Models
 * @module grease/config/models
 */

export { default as GreaseConfig } from './grease-config.model'
