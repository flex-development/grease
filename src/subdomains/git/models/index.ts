/**
 * @file Entry Point - Models
 * @module grease/git/models
 */

export { default as CommitGrammar } from './commit-grammar.model'
export { default as Commit } from './commit.model'
export { default as Repository } from './repository.model'
