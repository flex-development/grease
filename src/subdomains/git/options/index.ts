/**
 * @file Entry Point - Options
 * @module grease/git/options
 */

export { default as CommitGrammarOptions } from './commit-grammar.options'
export { default as GitCommitOptions } from './git-commit.options'
export { default as GitTagOptions } from './git-tag.options'
export { default as GitOptions } from './git.options'
