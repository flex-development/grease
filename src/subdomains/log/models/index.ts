/**
 * @file Entry Point - Models
 * @module grease/log/models
 */

export { default as LogObject } from './log-object.model'
