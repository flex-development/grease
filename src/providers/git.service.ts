/**
 * @file Providers - GitService
 * @module grease/providers/GitService
 */

import type { Commit } from '#src/interfaces'
import { CommitGrammar } from '#src/models'
import { CommitOptions, GitTagOptions } from '#src/options'
import type { BreakingChange, CommitLogField, Trailer } from '#src/types'
import {
  at,
  cast,
  construct,
  fallback,
  fork,
  get,
  ifelse,
  includes,
  isNaN,
  join,
  ksort,
  objectify,
  pick,
  select,
  sift,
  split,
  template,
  trim,
  trimEnd,
  type Construct
} from '@flex-development/tutils'
import { Injectable } from '@nestjs/common'
import consola from 'consola'
import { exec, type ExecOptions } from 'node:child_process'
import util from 'node:util'
import semver from 'semver'
import ValidationService from './validation.service'

/**
 * Git operations provider.
 *
 * @see https://git-scm.com/docs
 *
 * @class
 */
@Injectable()
class GitService {
  /**
   * Commit log format.
   *
   * @see https://git-scm.com/docs/git-log#_pretty_formats
   *
   * @public
   * @static
   * @readonly
   * @member {string} RAW_COMMIT_FORMAT
   */
  public static readonly RAW_COMMIT_FORMAT: string = join(
    [
      '%s%n',
      'author.email',
      '%n%ae%n',
      'author.name',
      '%n%an%n',
      'body',
      '%n%b%n',
      'date',
      '%n%cI%n',
      'hash',
      '%n%h%n',
      'sha',
      '%n%H%n',
      'tags',
      '%n%D%n',
      'trailers',
      '%n%(trailers)%n'
    ],
    '-'
  )

  /**
   * Create a git operations service.
   *
   * @param {ValidationService} validator - Validation service
   */
  constructor(protected readonly validator: ValidationService) {}

  /**
   * Get an array of parsed commits.
   *
   * @see {@linkcode CommitOptions}
   *
   * @public
   * @async
   *
   * @param {Partial<CommitOptions>?} [opts] - Commit options
   * @return {Promise<Commit[]>} Parsed commit array
   */
  public async commits(opts?: Partial<CommitOptions>): Promise<Commit[]> {
    const {
      cwd,
      debug,
      from,
      issue_prefixes,
      to
    } = await this.validator.validate(new CommitOptions(opts))

    /**
     * String used to separate commit logs.
     *
     * @const {string} LOG_DELIMITER
     */
    const LOG_DELIMITER: string = '--$--'

    /**
     * Git command arguments.
     *
     * @const {string[]} args
     */
    const args: string[] = sift([
      '--decorate-refs=refs/tags',
      '--decorate=short',
      template('--format=\'{format}{delimiter}\'', {
        delimiter: LOG_DELIMITER,
        format: GitService.RAW_COMMIT_FORMAT
      }),
      join(sift([from, to]), '..'),
      ifelse(cwd === process.cwd(), '', `-- ${cwd}`)
    ])

    /**
     * Raw commit logs.
     *
     * @const {{ stderr: string; stdout: string }} logs
     */
    const logs: { stderr: string; stdout: string } = await this.log(args, {
      cwd,
      debug
    })

    /**
     * Array containing parsed commits.
     *
     * @const {Commit[]} commits
     */
    const commits: Commit[] = []

    /**
     * Commit parser grammar.
     *
     * @const {CommitGrammar} grammar
     */
    const grammar: CommitGrammar = new CommitGrammar()

    // parse raw commits
    for (let chunk of split(logs.stdout, LOG_DELIMITER + '\n')) {
      if (!(chunk = trim(chunk))) continue

      // get commit header and raw fields
      const [[header = ''], raw_fields]: [string[], string[]] = fork(
        split(chunk, /^(?=-.*?-\n*$)/gm),
        raw => !grammar.field.test(raw)
      )

      // extract commit header data
      const {
        groups: {
          breaking = null,
          pr = 'null',
          scope = null,
          subject = '',
          type = ''
        } = {}
      } = pick(grammar.header.exec(header)!, ['groups'])

      /**
       * Commit fields object.
       *
       * @const {Construct<Record<CommitLogField, string>>} fields
       */
      const fields: Construct<Record<CommitLogField, string>> = construct(
        cast<Record<CommitLogField, string>>(
          objectify(
            raw_fields,
            raw => get(grammar.field.exec(raw)!.groups, 'field', ''),
            raw => trim(raw.replace(grammar.field, ''))
          )
        )
      )

      /**
       * Tags associated with commit.
       *
       * @const {string[]} tags
       */
      const tags: string[] = select(
        split(fields.tags, ','),
        tag => !!trim(tag),
        tag => trim(tag.replace(/^tag: */, ''))
      )

      /**
       * Commit message trailers.
       *
       * @const {Trailer[]} trailers
       */
      const trailers: Trailer[] = select(
        [...fields.trailers.matchAll(grammar.trailer)],
        null,
        match => pick(match.groups!, ['token', 'value'])
      )

      /**
       * Breaking changes noted in commit subject and trailers.
       *
       * @const {BreakingChange[]} breaking_changes
       */
      const breaking_changes: BreakingChange[] = select(
        trailers,
        trailer => /^BREAKING[ -]CHANGE/.test(trailer.token),
        trailer => ({ scope: null, subject: trailer.value, type })
      )

      // add subject to breaking changes if breaking change is in subject
      if (breaking && !includes(breaking_changes, subject)) {
        breaking_changes.unshift({ scope, subject, type })
      }

      /**
       * Parsed commit object.
       *
       * @var {Commit} commit
       */
      const commit: Commit = {
        ...fields,
        body: trimEnd(fields.body.replace(fields.trailers, '')),
        breaking: !!breaking_changes.length,
        breaking_changes,
        header: header.replace(/\n$/, ''),
        mentions: select(
          [...chunk.matchAll(grammar.mention)],
          null,
          match => get(match, 'groups.user')!
        ),
        pr: fallback(+pr, null, isNaN),
        references: select(
          [...chunk.matchAll(grammar.reference(issue_prefixes))],
          null,
          match => ({
            action: get(match, 'groups.action', null),
            number: +get(match, 'groups.number', ''),
            owner: get(match, 'groups.owner', null),
            ref: get(match, 'groups.ref', ''),
            repo: get(match, 'groups.repo', null)
          })
        ),
        scope,
        subject,
        tags,
        trailers,
        type,
        version: at(tags, 0, null)
      }

      commits.push(ksort(commit, { deep: true }))
    }

    return commits
  }

  /**
   * Get commit logs.
   *
   * @see https://git-scm.com/docs/git-log
   *
   * @public
   * @async
   *
   * @param {ReadonlyArray<string>} args - `git log` options
   * @param {ExecOptions & { debug?: boolean }} [opts={}] - Exec options
   * @return {Promise<{ stderr: string; stdout: string }>} Command output
   */
  public async log(
    args: readonly string[],
    opts: ExecOptions & { debug?: boolean } = {}
  ): Promise<{ stderr: string; stdout: string }> {
    /**
     * Command to execute.
     *
     * @const {string} command
     */
    const command: string = join(['git', 'log', ...args], ' ')

    // debug git log command
    if (opts.debug) consola.info(`[GitService.${this.log.name}]`, command)

    return util.promisify(exec)(command, {
      ...opts,
      encoding: 'utf8',
      maxBuffer: Number.POSITIVE_INFINITY,
      shell: process.env.SHELL
    })
  }

  /**
   * Get an array containing all git tags in reverse chronological order.
   *
   * @see {@linkcode GitTagOptions}
   *
   * @public
   * @async
   *
   * @param {Partial<GitTagOptions>} [opts] - Git tag options
   * @return {Promise<string[]>} Git tags array
   */
  public async tags(opts?: Partial<GitTagOptions>): Promise<string[]> {
    const {
      cwd,
      debug,
      tagprefix,
      unstable
    } = await this.validator.validate(new GitTagOptions(opts))

    /**
     * Raw commit logs.
     *
     * @const {{ stderr: string; stdout: string }} logs
     */
    const logs: { stderr: string; stdout: string } = await this.log([
      '--decorate-refs=refs/tags',
      '--decorate=short',
      '--format=%D'
    ], { cwd, debug, env: process.env })

    return select(
      select(
        split(logs.stdout, '\n'),
        tag => trim(tag).startsWith(`tag: ${tagprefix}`),
        tag => trim(tag).replace(/^tag: */, '')
      ),
      tag => (
        !!semver.valid(tag = tag.replace(new RegExp(`^${tagprefix}`), '')) &&
        (unstable ? true : !semver.parse(tag)!.prerelease.length)
      )
    )
  }
}

export default GitService
