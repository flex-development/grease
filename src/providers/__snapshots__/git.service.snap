// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:providers/GitService > #commits > should return parsed commit array 1`] = `
[
  {
    "author": {
      "email": "unicornware@flexdevelopment.llc",
      "name": "Lex",
    },
    "body": "",
    "breaking": false,
    "breaking_changes": [],
    "date": "2023-09-09T23:59:31+00:00",
    "hash": "a399eae",
    "header": "release: 1.0.0-alpha.23 (#361)",
    "mentions": [],
    "pr": 361,
    "references": [
      {
        "action": null,
        "number": 361,
        "owner": null,
        "ref": "#361",
        "repo": null,
      },
    ],
    "scope": null,
    "sha": "a399eaeff03a88cdb1d59fc1fb42d88fcdc773fe",
    "subject": "1.0.0-alpha.23 (#361)",
    "tags": [
      "1.0.0-alpha.23",
    ],
    "trailers": [
      {
        "token": "Signed-off-by",
        "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
      },
    ],
    "type": "release",
    "version": "1.0.0-alpha.23",
  },
  {
    "author": {
      "email": "unicornware@flexdevelopment.llc",
      "name": "Lexus Drumgold",
    },
    "body": "- prettier 3.0 formatting has degraded in quality for js-like files, but the team refuses to fix it
- prettier can be removed completely once dprint has its own yaml plugin + better json5 support
- prettier markdown formatting was always subpar; it never played nicely with markdownlint
- prettier/prettier#15358
- prettier/prettier#5715
- prettier/prettier#11881
- dprint/dprint#736
- dprint/dprint-plugin-typescript#432",
    "breaking": false,
    "breaking_changes": [],
    "date": "2023-09-09T19:00:25-04:00",
    "hash": "7f578c9",
    "header": "chore: dprint migration",
    "mentions": [],
    "pr": null,
    "references": [
      {
        "action": null,
        "number": 15358,
        "owner": "prettier",
        "ref": "#15358",
        "repo": "prettier",
      },
      {
        "action": null,
        "number": 5715,
        "owner": "prettier",
        "ref": "#5715",
        "repo": "prettier",
      },
      {
        "action": null,
        "number": 11881,
        "owner": "prettier",
        "ref": "#11881",
        "repo": "prettier",
      },
      {
        "action": null,
        "number": 736,
        "owner": "dprint",
        "ref": "#736",
        "repo": "dprint",
      },
      {
        "action": null,
        "number": 432,
        "owner": "dprint",
        "ref": "#432",
        "repo": "dprint-plugin-typescript",
      },
    ],
    "scope": null,
    "sha": "7f578c90c69d12b283e49b9036c06029585630f8",
    "subject": "dprint migration",
    "tags": [],
    "trailers": [
      {
        "token": "Signed-off-by",
        "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
      },
    ],
    "type": "chore",
    "version": null,
  },
  {
    "author": {
      "email": "unicornware@flexdevelopment.llc",
      "name": "Lexus Drumgold",
    },
    "body": "- https://github.com/flex-development/nest-commander/releases/tag/1.0.0-alpha.1",
    "breaking": false,
    "breaking_changes": [],
    "date": "2023-09-09T06:16:46-04:00",
    "hash": "6624182",
    "header": "build(deps-dev): use @flex-development/nest-commander in lieu of nest-commander",
    "mentions": [],
    "pr": null,
    "references": [],
    "scope": "deps-dev",
    "sha": "6624182ccc862dd2a690225372fb825b8c4b6714",
    "subject": "use @flex-development/nest-commander in lieu of nest-commander",
    "tags": [],
    "trailers": [
      {
        "token": "Signed-off-by",
        "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
      },
    ],
    "type": "build",
    "version": null,
  },
  {
    "author": {
      "email": "unicornware@flexdevelopment.llc",
      "name": "Lexus Drumgold",
    },
    "body": "",
    "breaking": false,
    "breaking_changes": [],
    "date": "2023-09-09T03:39:03-04:00",
    "hash": "978ffe9",
    "header": "fix(plugins): [\`decorators\`] decorator check",
    "mentions": [],
    "pr": null,
    "references": [],
    "scope": "plugins",
    "sha": "978ffe9ec9abfb25868f6ff547cfd45c87c4111a",
    "subject": "[\`decorators\`] decorator check",
    "tags": [],
    "trailers": [
      {
        "token": "Signed-off-by",
        "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
      },
    ],
    "type": "fix",
    "version": null,
  },
  {
    "author": {
      "email": "49699333+dependabot[bot]@users.noreply.github.com",
      "name": "dependabot[bot]",
    },
    "body": "",
    "breaking": false,
    "breaking_changes": [],
    "date": "2023-08-15T04:19:29+00:00",
    "hash": "613a69a",
    "header": "build(deps): Bump the flex-development group with 3 updates (#330)",
    "mentions": [],
    "pr": 330,
    "references": [
      {
        "action": null,
        "number": 330,
        "owner": null,
        "ref": "#330",
        "repo": null,
      },
    ],
    "scope": "deps",
    "sha": "613a69aae2ce4087e0e6a4eebfdc3b535bcf88b8",
    "subject": "Bump the flex-development group with 3 updates (#330)",
    "tags": [],
    "trailers": [
      {
        "token": "Signed-off-by",
        "value": "dependabot[bot] <support@github.com>",
      },
      {
        "token": "Co-authored-by",
        "value": "dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>",
      },
    ],
    "type": "build",
    "version": null,
  },
  {
    "author": {
      "email": "unicornware@flexdevelopment.llc",
      "name": "Lexus Drumgold",
    },
    "body": "",
    "breaking": true,
    "breaking_changes": [
      {
        "scope": "utils",
        "subject": "[\`analyzeOutputs\`] intake \`esbuild.Metafile['outputs']\`",
        "type": "refactor",
      },
    ],
    "date": "2023-04-07T19:38:09-04:00",
    "hash": "f467b4b",
    "header": "refactor(utils)!: [\`analyzeOutputs\`] intake \`esbuild.Metafile['outputs']\`",
    "mentions": [],
    "pr": null,
    "references": [],
    "scope": "utils",
    "sha": "f467b4bae340a83d126c8dc80b75af0a912d7ff7",
    "subject": "[\`analyzeOutputs\`] intake \`esbuild.Metafile['outputs']\`",
    "tags": [],
    "trailers": [
      {
        "token": "Signed-off-by",
        "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
      },
    ],
    "type": "refactor",
    "version": null,
  },
  {
    "author": {
      "email": "unicornware@flexdevelopment.llc",
      "name": "Lexus Drumgold",
    },
    "body": "- https://esbuild.github.io/api/#watch
- replaced \`esbuilder\` with \`createContext\` to support watch mode (and in the future, serve + live reload)
- started adding cli flags
- overhauled interfaces",
    "breaking": true,
    "breaking_changes": [
      {
        "scope": null,
        "subject": "watch mode",
        "type": "feat",
      },
    ],
    "date": "2023-03-31T02:50:34-04:00",
    "hash": "f1decde",
    "header": "feat!: watch mode",
    "mentions": [],
    "pr": null,
    "references": [],
    "scope": null,
    "sha": "f1decde8597173569c69f8df8b013559231ff18b",
    "subject": "watch mode",
    "tags": [],
    "trailers": [
      {
        "token": "Signed-off-by",
        "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
      },
    ],
    "type": "feat",
    "version": null,
  },
  {
    "author": {
      "email": "unicornware@flexdevelopment.llc",
      "name": "Lexus Drumgold",
    },
    "body": "- import.meta.url may be used in build config (e.g. evanw/esbuild#1921)
- https://github.com/unjs/mlly/blob/v0.5.14/src/eval.ts#L17-L38",
    "breaking": false,
    "breaking_changes": [],
    "date": "2022-09-27T03:23:45-04:00",
    "hash": "3e9c7c8",
    "header": "fix(config): prevent \`import.meta.url\` from being rewritten",
    "mentions": [],
    "pr": null,
    "references": [
      {
        "action": null,
        "number": 1921,
        "owner": "evanw",
        "ref": "#1921",
        "repo": "esbuild",
      },
    ],
    "scope": "config",
    "sha": "3e9c7c881cbc93156fe31176c7a0090e698dfac6",
    "subject": "prevent \`import.meta.url\` from being rewritten",
    "tags": [],
    "trailers": [
      {
        "token": "Signed-off-by",
        "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
      },
    ],
    "type": "fix",
    "version": null,
  },
  {
    "author": {
      "email": "unicornware@flexdevelopment.llc",
      "name": "Lex",
    },
    "body": "",
    "breaking": false,
    "breaking_changes": [],
    "date": "2022-09-15T21:22:09-04:00",
    "hash": "1601d26",
    "header": "feat: file-to-file transpilation (#4)",
    "mentions": [],
    "pr": 4,
    "references": [
      {
        "action": null,
        "number": 4,
        "owner": null,
        "ref": "#4",
        "repo": null,
      },
    ],
    "scope": null,
    "sha": "1601d268bce74abafab6a4b0d0e244b35dd1e808",
    "subject": "file-to-file transpilation (#4)",
    "tags": [],
    "trailers": [
      {
        "token": "Signed-off-by",
        "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
      },
    ],
    "type": "feat",
    "version": null,
  },
  {
    "author": {
      "email": "unicornware@flexdevelopment.llc",
      "name": "Lexus Drumgold",
    },
    "body": "",
    "breaking": false,
    "breaking_changes": [],
    "date": "2022-09-01T00:37:08-04:00",
    "hash": "9c58d1c",
    "header": "ci: add @dependabot config",
    "mentions": [
      "dependabot",
    ],
    "pr": null,
    "references": [],
    "scope": null,
    "sha": "9c58d1ca753406bbba92c729360b9b6e400a6e9a",
    "subject": "add @dependabot config",
    "tags": [],
    "trailers": [
      {
        "token": "Signed-off-by",
        "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
      },
    ],
    "type": "ci",
    "version": null,
  },
  {
    "author": {
      "email": "unicornware@flexdevelopment.llc",
      "name": "Lexus Drumgold",
    },
    "body": "",
    "breaking": false,
    "breaking_changes": [],
    "date": "2022-09-01T00:34:48-04:00",
    "hash": "50385b6",
    "header": "initial commit",
    "mentions": [],
    "pr": null,
    "references": [],
    "scope": null,
    "sha": "50385b68593d7b07f269631af0ca551539fe22d0",
    "subject": "",
    "tags": [],
    "trailers": [
      {
        "token": "Signed-off-by",
        "value": "Lexus Drumgold <unicornware@flexdevelopment.llc>",
      },
    ],
    "type": "",
    "version": null,
  },
]
`;

exports[`unit:providers/GitService > #tags > should return tags array with unstable tags 1`] = `
[
  "1.0.0-alpha.22",
  "1.0.0-alpha.21",
  "1.0.0-alpha.20",
  "1.0.0-alpha.19",
  "1.0.0-alpha.18",
  "1.0.0-alpha.17",
  "1.0.0-alpha.16",
  "1.0.0-alpha.15",
  "1.0.0-alpha.14",
  "1.0.0-alpha.13",
  "1.0.0-alpha.12",
  "1.0.0-alpha.11",
  "1.0.0-alpha.10",
  "1.0.0-alpha.9",
  "1.0.0-alpha.8",
  "1.0.0-alpha.7",
  "1.0.0-alpha.6",
  "1.0.0-alpha.5",
  "1.0.0-alpha.4",
  "1.0.0-alpha.3",
  "1.0.0-alpha.2",
  "1.0.0-alpha.1",
]
`;

exports[`unit:providers/GitService > #tags > should return tags array without unstable tags 1`] = `
[
  "tutils@5.0.1",
  "tutils@5.0.0",
  "tutils@4.8.0",
  "tutils@4.7.0",
  "tutils@4.6.0",
  "tutils@4.5.0",
  "tutils@4.4.0",
  "tutils@4.3.0",
  "tutils@4.2.3",
  "tutils@4.2.2",
  "tutils@4.2.1",
  "tutils@4.2.0",
  "tutils@4.1.1",
  "tutils@4.1.0",
  "tutils@4.0.3",
  "tutils@4.0.2",
  "tutils@4.0.1",
  "tutils@4.0.0",
  "tutils@3.1.7",
  "tutils@3.1.6",
  "tutils@3.1.5",
  "tutils@3.1.4",
  "tutils@3.1.3",
  "tutils@3.1.2",
  "tutils@3.1.1",
  "tutils@3.1.0",
  "tutils@3.0.0",
  "tutils@2.0.0",
  "tutils@1.0.0",
]
`;
